### ISSUE TITLE
Multi-Unused Register ROP Sets

### ISSUE THREAD

**Issue by bannsec** [Gallopsled/pwntools#937]
I find myself running into quite a few situations with ROP and pwntools where it complains about not being able to set a reg (for instance, rsi). Looking into the gadgets that pwntools already knows about, there has become a tendency for me to find something like the following:

```asm
pop rsi;
pop r15;
ret;
```

In this case, r15 is not a registers that matters or is used for that particular ROP call. However, it is being thrown out of possibly being an option because it isn't directly a pop rsi. The idea would be to allow find_gadgets (or setRegisters?) to find and return all gadgets that possibly have a pop for the register in question. Then the builder could simply throw out those gadgets that have a dependency on a different register we want to set (i.e.: the registers dict). Finally we could sort the remaining registers by number of register pops to return to ensure we're not adding length the the chain where we don't have to.

This could be improved on with other actual solving techniques, but i think as a naive technique for improving the ability to set registers and build chains it would be a useful step.

**Comment by zachriggle** [comment]
I don't plan to invest any time into making the `ROP` module more robust that is not part of #686

**Comment by zachriggle** [comment]
This should be fixed since #1019.

```
>>> context.arch='amd64'
>>> e=ELF.from_assembly('pop rsi; pop r15; pop rdi; ret')
[*] '/tmp/pwntools/pwn-asm-SV1jAZ/step3'
    Arch:     amd64-64-little
    RELRO:    No RELRO
    Stack:    No canary found
    NX:       NX disabled
    PIE:      No PIE (0x10000000)
    RWX:      Has RWX segments
>>> r=ROP(e)
[*] Loaded cached gadgets for '/tmp/pwntools/pwn-asm-SV1jAZ/step3'
>>> r.call(0xdeadbeef, [1,2])
>>> print r.dump()
0x0000:       0x10000000 pop rsi; pop r15; pop rdi; ret
0x0008:              0x2 [arg1] rsi = 2
0x0010:       'eaaafaaa' <pad r15>
0x0018:              0x1 [arg0] rdi = 1
0x0020:       0xdeadbeef
```
